File: 2022/05/p.py
#!/usr/bin/env python
import sys
import re

drawing = []
for line in sys.stdin:
    line = line.rstrip()
    if line == '':
        break
    else:
        drawing.append(line)

N = int(drawing.pop().split()[-1])
stacks = []
for i in range(N):
    stacks.append([])
    for line in drawing:
        j = 1+i*4
        if j < len(line) and line[j] != ' ':
            stacks[i].append(line[j])
    stacks[i] = list(reversed(stacks[i]))

stacks1 = [list(s) for s in stacks]
stacks2 = [list(s) for s in stacks]

for line in sys.stdin:
    m = re.fullmatch('move ([0-9]+) from ([0-9]) to ([0-9])$', line.strip())
    assert m, line
    k = int(m[1])
    i = int(m[2])
    j = int(m[3])

    for _ in range(k):
        stacks1[j-1].append(stacks1[i-1].pop())

    stacks2[j-1] += stacks2[i-1][-k:]
    stacks2[i-1] = stacks2[i-1][:-k]

ans1 = ''.join(s[-1] for s in stacks1)
ans2 = ''.join(s[-1] for s in stacks2)
print(ans1, ans2)

File: 2022/05/p.js
#!/usr/bin/env node
const fs = require('fs');

const text = fs.readFileSync(0).toString();

function parseDrawing(drawing) {
  const lines = drawing.split('\n');
  const n = lines[lines.length - 1].trim().split(/\s+/).length;
  let stacks = Array(n).fill().map(_ => []);
  for (let i = lines.length - 2; i >= 0; i--) {
    for (let j = 1, k = 0; j < lines[i].length; j += 4, k++) {
      if (lines[i][j] != ' ') {
        stacks[k].push(lines[i][j]);
      }
    }
  }
  return stacks;
}

function solve(part) {
  const [drawing, instr] = text.split('\n\n');
  let stacks = parseDrawing(drawing);
  for (const line of instr.split('\n')) {
    const m = line.match(/^move (\d+) from (\d) to (\d)$/);
    const k = parseInt(m[1]);
    const i = parseInt(m[2]) - 1;
    const j = parseInt(m[3]) - 1;
    if (part == 1) {
      for (let r = 0; r < k; r++) {
        stacks[j].push(stacks[i].pop());
      }
    } else {
      let els = stacks[i].splice(stacks[i].length - k);
      stacks[j] = stacks[j].concat(els);
    }
  }
  return stacks.map(s => s[s.length-1]).join('');
}

console.log(solve(1), solve(2));
